### ISSUE TITLE
Provide API to get all available versions for a project

### ISSUE THREAD

**Issue by martinklepsch** [cljdoc/cljdoc#350]
[Dash](https://www.kapeli.com/dash) has been using Clojars' API to figure out what versions are available (In Dash you can install documentation for older versions). 

E.g. https://clojars.org/api/artifacts/cljdoc/cljdoc

The issue is that we now also host documentation for various artifacts coming from Maven Central (core.async, Clojure itself, etc.) and for these artifacts this API doesn't work.

**We will need to provide our own endpoint to return the versions for those artifacts. The data returned can be based on our internal database.**

Perhaps a good approach would be to extend the `index-routes` below to handle additional content types (JSON):
```clj
(defn index-routes []
  #{["/versions" :get nop :route-name :cljdoc/index]
    ["/versions/:group-id" :get nop :route-name :group/index]
    ["/versions/:group-id/:artifact-id" :get nop :route-name :artifact/index]})
```

cc @holyjak feel free to add to this if you have any further thoughts.

**Comment by holyjak** [comment]
Can you show an example of the output Dash expects?

**Comment by holyjak** [comment]
@martinklepsch I assume the DB currently only has versions of clojars artifacts and not of Maven Central artifacts, correct? So we would need to extend some batch job or something to fetch and store those as well... What are your thoughts / recommendations?

**Comment by martinklepsch** [comment]
> I assume the DB currently only has versions of clojars artifacts and not of Maven Central artifacts, correct?

The `versions` table will only contain artifacts for which we have built documentation. So the `versions` table won't necessarily have Maven Central artifacts. 

I think what we could use however, and what would be close to the Clojars API that Dash is currently using is our Lucene index. The index contains stuff from Clojars and Maven Central and we can return the list of versions. One issue with this approach is that it might list artifacts for which we have not yet built documentation but that has also been true for the Clojars API. It might even be that Dash requests builds if documentation isn't available, can't quite remember right now. 

Do you see any issues with using the Lucene index for this? 

> Bogdan needs them sorted by release date though I guess most of the time sorting by version would work too

I think sorting by release date isn't really a thing. Maven really only cares about version numbers and the release dates are metadata that don't affect dependency resolution. So let's just sort them using `version-clj` like we already do on some index pages. 

> BTW is there any particular order to the entries returned by `sqlite/get-documented-versions`? It doesn't seem so from the SQL itself.

Not sorted in any way, I think SQL itself isn't able to sort with the necessary logic (compared to `version-clj`). We could process the result after receiving the results though. Didn't seem necessary yet I guess :)

**Comment by Kapeli** [comment]
Sorted by version is what I want, sorry. No idea why I said sorted by release date.

**Comment by holyjak** [comment]
@martinklepsch Why would the search index be better than the `versions` table? Because it has versions that we have not build docs for yet and because Dash _might_ be able to request them to be build? (If Dash cannot do that then I see no point in showing a version that has no docset to it.)

An advantage of the `versions` table is that it knows of (built) older versions of Maven Artifacts, while the index only contains the latest version for Maven Central (We'd need a separate query for each artifact to fetch its past versions, which we currently don't do.)

Here is a draft of the implementation using the table: #357 I haven't tested it yet but would like to hear you thoughts on the approach (and see my 2 review comments).

**Comment by holyjak** [comment]
@Kapeli Is it OK if the response has the same format for all the  request granularities (only being more filtered), i.e. a tree of `<group> -> <artifact> -> <versions>`?. Example, with 2 indexed artifacts:

```
/versions => {"clj-concordion":{"clj-concordion":["1.0.7"]},"metosin":{"muuntaja":["0.6.5","0.6.0"]}}
/versions/metosin => {"metosin":{"muuntaja":["0.6.5","0.6.0"]}}
/versions/metosin/muuntaja => {"metosin":{"muuntaja":["0.6.5","0.6.0"]}}
```

**Comment by Kapeli** [comment]
Sure, that's ok. Thank you!

**Comment by martinklepsch** [comment]
@Kapeli we just deployed this change, you can use it like this:

```sh
curl -H 'accept: application/json' https://cljdoc.org/versions | jq   # large payload!
curl -H 'accept: application/json' https://cljdoc.org/versions/org.clojure | jq
curl -H 'accept: application/json' https://cljdoc.org/versions/org.clojure/clojure | jq
```

Thanks again to @holyjak for leading this, excited to have non Clojars packages properly available in cljdoc! 🚀

**Comment by Kapeli** [comment]
Awesome. I'll update Dash ASAP.

**Comment by Kapeli** [comment]
Sorry for taking so long. 

I'm working on adding support for this, but it seems this API only returns the versions for which cljdoc.org has already built docs for.

Dash is meant to request doc builds from cljdoc.org when cljdoc.org doesn't have the needed version already built. If I switch to using this API, users will just get an error when trying to install a package for which there are no docs yet, instead of Dash requesting the docs being built as it does now.

**Comment by Kapeli** [comment]
I just noticed you were discussing whether or not to include unbuilt versions as well. I completely missed those comments, sorry! I was only reading the ones mentioning @Kapeli ☹️

**Comment by holyjak** [comment]
@Kapeli a new release that returns all, even unbuilt versions is being deployed (or do you need to do that manually, @martinklepsch ?).

**Comment by martinklepsch** [comment]
Changes are being deployed as they are pushed to `master` :)

**Comment by holyjak** [comment]
@Kapeli Has Dash been updated yet to use the new API? Thank you!

**Comment by Kapeli** [comment]
I'm waiting for https://github.com/cljdoc/cljdoc/issues/350#issuecomment-557346556 to be addressed.

For example, https://cljdoc.org/versions/fast-zip-visit returns 0 versions, but https://cljdoc.org/d/fast-zip-visit/fast-zip-visit/1.0.2 is available (although unbuilt).

**Comment by holyjak** [comment]
Sorry, bad communication on my part. It is fixed, you just need to append `all=true` to the request: https://cljdoc.org/versions/fast-zip-visit?all=true

**Comment by Kapeli** [comment]
Great. The change will happen in the next Dash update (v5.1.0). I should release it in a few days.

**Comment by Kapeli** [comment]
I've just released Dash 5.1.0 🎉
